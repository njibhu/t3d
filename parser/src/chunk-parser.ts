import { BaseType, DataType } from "./types";

interface Definition {
  chunkName: string;
  name: string;
  version: number;
  definitions: {
    [definition: string]: { [key: string]: DataType | string };
  };
  root: {
    [key: string]: DataType | string;
  };
}

interface ParseFunctionReturn {
  newPosition: number;
  data: any;
}

export class ChunkParser implements Definition {
  public readonly chunkName: Definition["chunkName"];
  public readonly name: Definition["name"];
  public readonly version: Definition["version"];
  public readonly definitions: Definition["definitions"];
  public readonly root: Definition["root"];

  constructor(definition: Definition) {
    Object.assign(this, definition);
  }

  public parse(dv: DataView, pos: number): ParseFunctionReturn {
    let position = pos;
    const parsedObject = {};
    for (const key in this.root) {
      const value = this.root[key];
      let parsedResult: ParseFunctionReturn;
      if (typeof value === "string") {
        parsedResult = this.parseType(dv, position, value);
      } else {
        const { baseType, subType, length } = value;
        parsedResult = this[baseType](dv, position, subType, length);
      }
      parsedObject[key] = parsedResult.data;
      position = parsedResult.newPosition;
    }

    return {
      newPosition: position,
      data: parsedObject,
    };
  }

  public parseType(dv: DataView, pos: number, typeDefinitionName: string): ParseFunctionReturn {
    const parsedObject = {};
    let position = pos;
    const definition = this.definitions[typeDefinitionName];
    for (const key in definition) {
      const value = definition[key];
      let parsedResult: ParseFunctionReturn;
      if (typeof value === "string") {
        parsedResult = this.parseType(dv, position, value);
      } else {
        const { baseType, subType, length } = value;
        parsedResult = this[baseType](dv, position, subType, length);
      }
      parsedObject[key] = parsedResult.data;

      position = parsedResult.newPosition;
    }

    return {
      newPosition: position,
      data: parsedObject,
    };
  }

  /**
   *      Parsers
   **/

  private Float32(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 4, data: dv.getFloat32(pos, true) };
  }

  private Float64(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 8, data: dv.getFloat64(pos, true) };
  }

  private Uint8(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 1, data: dv.getUint8(pos) };
  }

  private Uint16(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 2, data: dv.getUint16(pos, true) };
  }

  private Uint32(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 4, data: dv.getUint32(pos, true) };
  }

  private Uint64(dv: DataView, pos: number): ParseFunctionReturn {
    return {
      newPosition: pos + 8,
      data: (BigInt(dv.getUint32(pos + 4, true)) << BigInt(32)) | BigInt(dv.getUint32(pos, true)),
    };
  }

  private CString(dv: DataView, pos: number, _subType: DataType | string, length: number): ParseFunctionReturn {
    const u8 = new Uint8Array(dv.buffer, pos);
    const end = length || u8.findIndex((v) => v === 0);

    return {
      newPosition: pos + end,
      data: String.fromCharCode.apply(null, new Uint8Array(u8.slice(0, end))),
    };
  }

  private FixedArray(dv: DataView, pos: number, type: DataType | string, length: number): ParseFunctionReturn {
    // Some types can be mapped directly from their buffer into the return type
    if (typeof type != "string" && this._getOptimisedArrayConstructor(type.baseType)) {
      return this.optimisedArray(dv, pos, type, length);
    }

    const data = [];
    let newPosition = pos;
    for (let itemIndex = 0; itemIndex < length; itemIndex++) {
      const parsedItem =
        typeof type === "string"
          ? this.parseType(dv, newPosition, type)
          : this[type.baseType](dv, newPosition, type.subType, type.length);
      data.push(parsedItem.data);
      newPosition = parsedItem.newPosition;
    }

    return {
      newPosition,
      data,
    };
  }

  private DynArray(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    let arrayLength = dv.getUint32(pos, true);
    let arrayOffset = dv.getUint32(pos + 4, true);

    if (arrayOffset === 0) {
      return {
        newPosition: pos + 8,
        data: [],
      };
    }
    let arrayPtr = pos + 4 + arrayOffset;

    return {
      newPosition: pos + 8,
      data: this.FixedArray(dv, arrayPtr, type, arrayLength).data,
    };
  }

  private RefArray(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    let data = [];
    let arrayLength = dv.getUint32(pos, true);
    let arrayPtr = dv.getUint32(pos + 4, true) + pos;
    if (arrayLength === 0) {
      return {
        data,
        newPosition: pos + 8,
      };
    }

    const pointer = dv.getUint32(pos + 4, true) + pos + 4;
    for (let index = 0; index < arrayLength; index++) {
      const offset = dv.getInt32(arrayPtr + 4 * index);
      if (offset !== 0) {
        let newPosition = pointer + index * 4 + offset;
        if (typeof type === "string") {
          data.push(this.parseType(dv, newPosition, type).data);
        } else {
          data.push(this[type.baseType](dv, newPosition, type.subType, type.length).data);
        }
      }
    }

    return {
      newPosition: pos + 8,
      data,
    };
  }

  private Pointer(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    const offset = dv.getUint32(pos, true);
    const parsedItem =
      typeof type === "string"
        ? this.parseType(dv, pos + offset, type)
        : this[type.baseType](dv, pos + offset, type.subType, type.length);

    return {
      newPosition: parsedItem.newPosition,
      data: parsedItem.data,
    };
  }

  private String16(dv: DataView, pos: number): ParseFunctionReturn {
    let newPosition = pos;
    let ptr = pos + dv.getUint32(pos, true);
    newPosition += 4;

    let data = "";
    let num;
    while (ptr + 2 < dv.byteLength && (num = dv.getUint16(ptr)) !== 0) {
      ptr += 2;
      data += String.fromCharCode(num);
    }

    return {
      newPosition,
      data,
    };
  }

  private Fileref(dv: DataView, pos: number): ParseFunctionReturn {
    return this.Filename(dv, pos);
  }

  private Filename(dv: DataView, pos: number): ParseFunctionReturn {
    // This implementation is based on the old Utils.getFileNameReader() function
    let newPosition = pos;
    try {
      let ptr = newPosition + dv.getUint32(newPosition, true);
      newPosition += 4;

      const m_lowPart = dv.getUint16(ptr);
      ptr += 2;
      const m_highPart = dv.getUint16(ptr);
      ptr += 2;
      const _m_terminator = dv.getUint16(ptr);
      ptr += 2;

      /// Getting the file name...
      /// Both need to be >= than 256 (terminator is 0)
      const ret = 0xff00 * (m_highPart - 0x100) + (m_lowPart - 0x100) + 1;

      return {
        newPosition,
        data: ret > 0 ? ret : 0,
      };
    } catch (e) {
      return {
        newPosition,
        data: -1,
      };
    }
  }

  private Unknown(dv: DataView, pos: number): ParseFunctionReturn {
    throw new Error("Could not parse unknown data");
  }

  /**
   *      Parser utils & helpers
   **/

  private optimisedArray(dv: DataView, pos: number, type: DataType, length: number): ParseFunctionReturn {
    const OptimisedArray = this._getOptimisedArrayConstructor(type.baseType);
    return {
      newPosition: pos + length * OptimisedArray.BYTES_PER_ELEMENT,
      data: Array.from(new OptimisedArray(dv.buffer, pos, length)),
    };
  }

  private _getOptimisedArrayConstructor(baseType: BaseType) {
    switch (baseType) {
      case BaseType.Float32:
        return Float32Array;

      case BaseType.Float64:
        return Float64Array;

      case BaseType.Uint8:
        return Uint8Array;

      case BaseType.Uint16:
        return Uint16Array;

      // case BaseType.Uint32:
      //   return Uint32Array;
    }
  }
}
