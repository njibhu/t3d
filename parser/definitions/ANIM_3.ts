import { Uint64, Fileref, Float32, Uint32, DynArray } from "../src/types";

export const V0 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV0",
  version: 0,
  definitions: {
    PackEmoteAnimationV0: {
      token: Uint64,
      timings: DynArray("PackEmoteTimingV0")
    },
    PackEmoteTimingV0: {
      modelFileId: Fileref(),
      blendInTime: Float32,
      blendOutTime: Float32,
      duration: Uint32
    }
  },
  root: {
    animations: DynArray("PackEmoteAnimationV0")
  }
};

export const V1 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV1",
  version: 1,
  definitions: {
    PackEmoteAnimationV1: {
      token: Uint64,
      timings: DynArray("PackEmoteTimingV1")
    },
    PackEmoteTimingV1: {
      modelFileId: Fileref(),
      blendInTime: Float32,
      blendOutTime: Float32,
      duration: Uint32,
      loopDuration: Uint32
    }
  },
  root: {
    animations: DynArray("PackEmoteAnimationV1")
  }
};

export const V2 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV2",
  version: 2,
  definitions: {
    PackEmoteAnimationV2: {
      Token: Uint64,
      Timing: DynArray("PackEmoteTimingV2")
    },
    PackEmoteTimingV2: {
      ModelFile: Fileref(),
      BlendIn: Float32,
      BlendOut: Float32,
      IntroDuration: Uint32,
      LoopDuration: Uint32,
      OutroDuration: Uint32
    }
  },
  root: {
    Animation: DynArray("PackEmoteAnimationV2")
  }
};

export const V3 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV3",
  version: 3,
  definitions: {
    PackEmoteAnimationV3: {
      Token: Uint64,
      Timing: DynArray("PackEmoteTimingV3")
    },
    PackEmoteTimingV3: {
      ModelFile: Fileref(),
      BlendIn: Float32,
      BlendOut: Float32,
      IntroDuration: Uint32,
      LoopDuration: Uint32,
      OutroDuration: Uint32,
      StartOffset: Uint32
    }
  },
  root: {
    Animation: DynArray("PackEmoteAnimationV3")
  }
};

export const latest = V3;
export const definitionArray = [V0, V1, V2, V3];