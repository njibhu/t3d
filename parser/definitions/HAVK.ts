import { FixedArray, Float32, Uint16, DynArray, Uint8, Uint64, Uint32 } from "../src/types";

export const V6 = {
  chunkName: "havk",
  name: "PackMapCollideV6",
  version: 6,
  definitions: {
    PackMapCollideMoppCodeV6: {
      cookedData: DynArray(Uint8)
    },
    PackMapCollideMeshV6: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      sizes: DynArray("PackMapCollideMeshSizeV6")
    },
    PackMapCollideMeshSizeV6: {
      scale: Float32,
      moppCodeIndex: Uint32
    },
    PackMapCollideGeometryV6: {
      meshRefs: DynArray("PackMapCollideMeshRefV6")
    },
    PackMapCollideMeshRefV6: {
      sequence: Uint64,
      meshIndex: Uint32
    },
    PackMapCollideModelPropV6: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV6: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    moppCodes: DynArray("PackMapCollideMoppCodeV6"),
    meshes: DynArray("PackMapCollideMeshV6"),
    geometries: DynArray("PackMapCollideGeometryV6"),
    propModels: DynArray("PackMapCollideModelPropV6"),
    zoneModels: DynArray("PackMapCollideModelZoneV6")
  }
};

export const V7 = {
  chunkName: "havk",
  name: "PackMapCollideV7",
  version: 7,
  definitions: {
    PackMapCollideMoppCodeV7: {
      cookedData: DynArray(Uint8)
    },
    PackMapCollideMeshV7: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      sizes: DynArray("PackMapCollideMeshSizeV7")
    },
    PackMapCollideMeshSizeV7: {
      scale: Float32,
      moppCodeIndex: Uint32
    },
    PackMapCollideGeometryV7: {
      meshRefs: DynArray("PackMapCollideMeshRefV7")
    },
    PackMapCollideMeshRefV7: {
      sequence: Uint64,
      meshIndex: Uint32
    },
    PackMapCollideModelPropV7: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV7: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideAiChunkV7: {
      navMeshData: DynArray(Uint8),
      coarseGraphData: DynArray(Uint8)
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    moppCodes: DynArray("PackMapCollideMoppCodeV7"),
    meshes: DynArray("PackMapCollideMeshV7"),
    geometries: DynArray("PackMapCollideGeometryV7"),
    propModels: DynArray("PackMapCollideModelPropV7"),
    zoneModels: DynArray("PackMapCollideModelZoneV7"),
    aiBoundaryMin: FixedArray(Float32, 3),
    aiBoundaryMax: FixedArray(Float32, 3),
    aiChunkDims: FixedArray(Uint32, 2),
    aiChunks: DynArray("PackMapCollideAiChunkV7")
  }
};

export const V8 = {
  chunkName: "havk",
  name: "PackMapCollideV8",
  version: 8,
  definitions: {
    PackMapCollideMeshV8: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      moppCodeScale: Float32,
      moppCodeData: DynArray(Uint8)
    },
    PackMapCollideGeometryV8: {
      meshRefs: DynArray("PackMapCollideMeshRefV8")
    },
    PackMapCollideMeshRefV8: {
      sequence: Uint64,
      meshIndex: Uint32
    },
    PackMapCollideModelPropV8: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV8: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideAiChunkV8: {
      navMeshData: DynArray(Uint8),
      coarseGraphData: DynArray(Uint8)
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    meshes: DynArray("PackMapCollideMeshV8"),
    geometries: DynArray("PackMapCollideGeometryV8"),
    propModels: DynArray("PackMapCollideModelPropV8"),
    zoneModels: DynArray("PackMapCollideModelZoneV8"),
    aiBoundaryMin: FixedArray(Float32, 3),
    aiBoundaryMax: FixedArray(Float32, 3),
    aiChunkDims: FixedArray(Uint32, 2),
    aiChunks: DynArray("PackMapCollideAiChunkV8")
  }
};

export const V9 = {
  chunkName: "havk",
  name: "PackMapCollideV9",
  version: 9,
  definitions: {
    PackMapCollideMeshV9: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      moppCodeScale: Float32,
      moppCodeData: DynArray(Uint8)
    },
    PackMapCollideGeometryV9: {
      meshRefs: DynArray("PackMapCollideMeshRefV9")
    },
    PackMapCollideMeshRefV9: {
      sequence: Uint64,
      meshIndex: Uint32
    },
    PackMapCollideModelObsV9: {
      geometryIndex: Uint32
    },
    PackMapCollideModelPropV9: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV9: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideAiChunkV9: {
      navMeshData: DynArray(Uint8),
      coarseGraphData: DynArray(Uint8)
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    meshes: DynArray("PackMapCollideMeshV9"),
    geometries: DynArray("PackMapCollideGeometryV9"),
    obsModels: DynArray("PackMapCollideModelObsV9"),
    propModels: DynArray("PackMapCollideModelPropV9"),
    zoneModels: DynArray("PackMapCollideModelZoneV9"),
    aiBoundaryMin: FixedArray(Float32, 3),
    aiBoundaryMax: FixedArray(Float32, 3),
    aiChunkDims: FixedArray(Uint32, 2),
    aiChunks: DynArray("PackMapCollideAiChunkV9")
  }
};

export const V10 = {
  chunkName: "havk",
  name: "PackMapCollideV10",
  version: 10,
  definitions: {
    PackMapCollideMeshV10: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      moppCodeScale: Float32,
      moppCodeData: DynArray(Uint8)
    },
    PackMapCollideGeometryV10: {
      meshRefs: DynArray("PackMapCollideMeshRefV10")
    },
    PackMapCollideMeshRefV10: {
      sequence: Uint64,
      meshIndex: Uint32
    },
    PackMapCollideModelObsV10: {
      geometryIndex: Uint32
    },
    PackMapCollideModelPropV10: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV10: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideAiChunkV10: {
      navMeshData: DynArray(Uint8),
      coarseGraphData: DynArray(Uint8),
      queryMediatorMoppData: DynArray(Uint8)
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    meshes: DynArray("PackMapCollideMeshV10"),
    geometries: DynArray("PackMapCollideGeometryV10"),
    obsModels: DynArray("PackMapCollideModelObsV10"),
    propModels: DynArray("PackMapCollideModelPropV10"),
    zoneModels: DynArray("PackMapCollideModelZoneV10"),
    aiBoundaryMin: FixedArray(Float32, 3),
    aiBoundaryMax: FixedArray(Float32, 3),
    aiChunkDims: FixedArray(Uint32, 2),
    aiChunks: DynArray("PackMapCollideAiChunkV10")
  }
};

export const V11 = {
  chunkName: "havk",
  name: "PackMapCollideV11",
  version: 11,
  definitions: {
    PackMapCollideMeshV11: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      moppCodeScale: Float32,
      moppCodeData: DynArray(Uint8)
    },
    PackMapCollideGeometryV11: {
      quantizedExtents: Uint8,
      meshRefs: DynArray("PackMapCollideMeshRefV11")
    },
    PackMapCollideMeshRefV11: {
      sequence: Uint64,
      meshIndex: Uint32
    },
    PackMapCollideModelObsV11: {
      translate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelPropV11: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV11: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideAiChunkV11: {
      boundsMin: FixedArray(Float32, 3),
      boundsMax: FixedArray(Float32, 3),
      navMeshData: DynArray(Uint8),
      coarseGraphData: DynArray(Uint8),
      queryMediatorMoppData: DynArray(Uint8)
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    meshes: DynArray("PackMapCollideMeshV11"),
    geometries: DynArray("PackMapCollideGeometryV11"),
    obsModels: DynArray("PackMapCollideModelObsV11"),
    propModels: DynArray("PackMapCollideModelPropV11"),
    zoneModels: DynArray("PackMapCollideModelZoneV11"),
    aiChunkDims: FixedArray(Uint32, 2),
    aiChunks: DynArray("PackMapCollideAiChunkV11")
  }
};

export const V12 = {
  chunkName: "havk",
  name: "PackMapCollideV12",
  version: 12,
  definitions: {
    PackMapCollideMeshV12: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      surfaces: DynArray(Uint16),
      moppCodeScale: Float32,
      moppCodeData: DynArray(Uint8)
    },
    PackMapCollideGeometryV12: {
      quantizedExtents: Uint8,
      meshRefs: DynArray("PackMapCollideMeshRefV12")
    },
    PackMapCollideMeshRefV12: {
      sequence: Uint64,
      meshIndex: Uint32
    },
    PackMapCollideModelObsV12: {
      translate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelPropV12: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV12: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideAiChunkV12: {
      boundsMin: FixedArray(Float32, 3),
      boundsMax: FixedArray(Float32, 3),
      navMeshData: DynArray(Uint8),
      coarseGraphData: DynArray(Uint8),
      queryMediatorMoppData: DynArray(Uint8)
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    meshes: DynArray("PackMapCollideMeshV12"),
    geometries: DynArray("PackMapCollideGeometryV12"),
    obsModels: DynArray("PackMapCollideModelObsV12"),
    propModels: DynArray("PackMapCollideModelPropV12"),
    zoneModels: DynArray("PackMapCollideModelZoneV12"),
    aiChunkDims: FixedArray(Uint32, 2),
    aiChunks: DynArray("PackMapCollideAiChunkV12")
  }
};

export const V13 = {
  chunkName: "havk",
  name: "PackMapCollideV13",
  version: 13,
  definitions: {
    PackMapCollideCollisionV13: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      surfaces: DynArray(Uint16),
      moppCodeData: DynArray(Uint8)
    },
    PackMapCollideBlockerV13: {
      vertices: DynArray(FixedArray(Float32, 3))
    },
    PackMapCollideAnimationV13: {
      sequence: Uint64,
      collisionIndices: DynArray(Uint32),
      blockerIndices: DynArray(Uint32)
    },
    PackMapCollideGeometryV13: {
      quantizedExtents: Uint8,
      animations: DynArray(Uint32)
    },
    PackMapCollideModelObsV13: {
      translate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelPropV13: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV13: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    collisions: DynArray("PackMapCollideCollisionV13"),
    blockers: DynArray("PackMapCollideBlockerV13"),
    animations: DynArray("PackMapCollideAnimationV13"),
    geometries: DynArray("PackMapCollideGeometryV13"),
    obsModels: DynArray("PackMapCollideModelObsV13"),
    propModels: DynArray("PackMapCollideModelPropV13"),
    zoneModels: DynArray("PackMapCollideModelZoneV13")
  }
};

export const V14 = {
  chunkName: "havk",
  name: "PackMapCollideV14",
  version: 14,
  definitions: {
    PackMapCollideCollisionV14: {
      indices: DynArray(Uint16),
      vertices: DynArray(FixedArray(Float32, 3)),
      surfaces: DynArray(Uint16),
      moppCodeData: "PackMoppType"
    },
    PackMoppType: {
      moppData: DynArray(Uint8)
    },
    PackMapCollideBlockerV14: {
      vertices: DynArray(FixedArray(Float32, 3))
    },
    PackMapCollideNavMeshV14: {
      navMesh: DynArray(Uint8),
      graph: DynArray(Uint8),
      mediator: DynArray(Uint8)
    },
    PackMapCollideAnimationV14: {
      sequence: Uint64,
      collisionIndices: DynArray(Uint32),
      blockerIndices: DynArray(Uint32)
    },
    PackMapCollideGeometryV14: {
      quantizedExtents: Uint8,
      animations: DynArray(Uint32),
      navMeshIndex: Uint16
    },
    PackMapCollideModelObsV14: {
      translate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelPropV14: {
      token: Uint64,
      sequence: Uint64,
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    },
    PackMapCollideModelZoneV14: {
      scale: Float32,
      translate: FixedArray(Float32, 3),
      rotate: FixedArray(Float32, 3),
      geometryIndex: Uint32
    }
  },
  root: {
    boundsMin: FixedArray(Float32, 3),
    boundsMax: FixedArray(Float32, 3),
    collisions: DynArray("PackMapCollideCollisionV14"),
    blockers: DynArray("PackMapCollideBlockerV14"),
    navMeshes: DynArray("PackMapCollideNavMeshV14"),
    animations: DynArray("PackMapCollideAnimationV14"),
    geometries: DynArray("PackMapCollideGeometryV14"),
    obsModels: DynArray("PackMapCollideModelObsV14"),
    propModels: DynArray("PackMapCollideModelPropV14"),
    zoneModels: DynArray("PackMapCollideModelZoneV14")
  }
};

export const latest = V14;
export const definitionArray = [V6, V7, V8, V9, V10, V11, V12, V13, V14];