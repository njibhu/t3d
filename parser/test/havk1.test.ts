import { latest as HAVKDef } from "../definitions/HAVK";
import { ChunkParser } from "../src/chunk-parser";
import { snapshot } from "./havk1";
import { transformSnapshot, toArrayBuffer } from "./test-helper";
import * as fs from "fs";

const havk1Result = transformSnapshot(snapshot);

test("matches for havk1", function () {
  const chunkBuffer = fs.readFileSync("./test/havk1.bin", null);
  const chunkParser = new ChunkParser(HAVKDef);

  const dv = new DataView(toArrayBuffer(chunkBuffer));
  const pos = 28; // After all the chunk and file headers

  const test = chunkParser.parse(dv, pos);
  expect(test.data).toMatchObject(havk1Result);
});
