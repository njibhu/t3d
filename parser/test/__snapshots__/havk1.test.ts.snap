// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches for havk1: havk1 1`] = `
{
  "animations": [
    {
      "blockerIndices": [],
      "collisionIndices": Uint32Array [
        0,
      ],
      "sequence": 21154994n,
    },
    {
      "blockerIndices": [],
      "collisionIndices": Uint32Array [
        0,
      ],
      "sequence": 135770290n,
    },
    {
      "blockerIndices": [],
      "collisionIndices": Uint32Array [
        0,
      ],
      "sequence": 175158308n,
    },
    {
      "blockerIndices": [],
      "collisionIndices": Uint32Array [
        0,
      ],
      "sequence": 272061484n,
    },
    {
      "blockerIndices": [],
      "collisionIndices": [],
      "sequence": 27366411428n,
    },
    {
      "blockerIndices": [],
      "collisionIndices": Uint32Array [
        0,
      ],
      "sequence": 192720390330n,
    },
    {
      "blockerIndices": [],
      "collisionIndices": Uint32Array [
        0,
      ],
      "sequence": 188074801256444082n,
    },
    {
      "blockerIndices": [],
      "collisionIndices": Uint32Array [
        0,
      ],
      "sequence": 188074801392735276n,
    },
  ],
  "blockers": [],
  "boundsMax": Float32Array [
    0,
    0,
    0,
  ],
  "boundsMin": Float32Array [
    0,
    0,
    0,
  ],
  "collisions": [
    {
      "indices": Uint16Array [
        0,
        2,
        1,
        3,
        0,
        1,
        3,
        1,
        4,
        5,
        4,
        1,
        5,
        1,
        2,
        5,
        2,
        6,
        7,
        6,
        2,
        0,
        7,
        2,
        8,
        7,
        0,
        8,
        0,
        3,
        8,
        3,
        9,
        9,
        3,
        4,
        9,
        4,
        10,
        5,
        10,
        4,
        5,
        6,
        10,
        11,
        10,
        6,
        7,
        11,
        6,
        8,
        11,
        7,
        8,
        9,
        11,
        11,
        9,
        10,
      ],
      "moppCodeData": {
        "moppData": Uint8Array [
          170,
          98,
          194,
          189,
          172,
          47,
          158,
          189,
          39,
          64,
          139,
          189,
          24,
          172,
          84,
          75,
          133,
          0,
          0,
          0,
          40,
          0,
          255,
          39,
          0,
          32,
          38,
          0,
          31,
          16,
          19,
          13,
          73,
          17,
          32,
          25,
          63,
          17,
          7,
          0,
          6,
          18,
          255,
          0,
          1,
          60,
          67,
          23,
          15,
          12,
          37,
          24,
          127,
          124,
          21,
          18,
          255,
          0,
          13,
          18,
          255,
          0,
          8,
          38,
          0,
          1,
          32,
          0,
          1,
          59,
          50,
          58,
          40,
          0,
          1,
          51,
          17,
          16,
          0,
          4,
          40,
          254,
          255,
          66,
          40,
          0,
          1,
          61,
          19,
          16,
          134,
          4,
          40,
          0,
          1,
          52,
          40,
          254,
          255,
          57,
          17,
          32,
          25,
          1,
          48,
          49,
          33,
          15,
          22,
          24,
          137,
          134,
          12,
          19,
          8,
          127,
          4,
          40,
          0,
          1,
          62,
          40,
          254,
          255,
          65,
          16,
          31,
          18,
          1,
          63,
          64,
          23,
          26,
          23,
          12,
          19,
          16,
          134,
          4,
          40,
          0,
          1,
          53,
          40,
          254,
          255,
          56,
          17,
          32,
          15,
          1,
          54,
          55,
          1,
        ],
      },
      "surfaces": Uint16Array [
        65535,
      ],
      "vertices": [
        Float32Array [
          -0.0066716657020151615,
          0.07288872450590134,
          1.1263327598571777,
        ],
        Float32Array [
          -0.0949147492647171,
          0.0442168153822422,
          -0.06799346208572388,
        ],
        Float32Array [
          -0.0066716657020151615,
          0.07288872450590134,
          -0.06799346208572388,
        ],
        Float32Array [
          -0.0949147492647171,
          0.0442168153822422,
          1.1263327598571777,
        ],
        Float32Array [
          -0.0949147567152977,
          -0.048567455261945724,
          -0.06799346208572388,
        ],
        Float32Array [
          -0.031061461195349693,
          -0.002175325993448496,
          -0.06799346208572388,
        ],
        Float32Array [
          0.04786554351449013,
          -0.002175325993448496,
          -0.06799346208572388,
        ],
        Float32Array [
          0.04786554351449013,
          -0.002175325993448496,
          1.1263327598571777,
        ],
        Float32Array [
          -0.031061461195349693,
          -0.002175325993448496,
          1.1263327598571777,
        ],
        Float32Array [
          -0.0949147567152977,
          -0.048567455261945724,
          1.1263327598571777,
        ],
        Float32Array [
          -0.006671677343547344,
          -0.07723937183618546,
          -0.06799346208572388,
        ],
        Float32Array [
          -0.006671677343547344,
          -0.07723937183618546,
          1.1263327598571777,
        ],
      ],
    },
  ],
  "geometries": [
    {
      "animations": Uint32Array [
        0,
        1,
        2,
        3,
        4,
        5,
        6,
        7,
      ],
      "navMeshIndex": 0,
      "quantizedExtents": 1,
    },
  ],
  "navMeshes": [],
  "obsModels": [],
  "propModels": [],
  "zoneModels": [],
}
`;
